lib/cretonne/src/isa/intel/mod.rs

```diff
@@ -111,4 +117,149 @@ impl TargetIsa for Isa {
     fn reloc_names(&self) -> &'static [&'static str] {
         &binemit::RELOC_NAMES
     }
+
+    fn prologue_epilogue(&self, func: &mut ir::Function) -> result::CtonResult {
+        let word_size = if self.flags().is_64bit() { 8 } else { 4 };
+        let csr_type = if self.flags().is_64bit() {
+            ir::types::I64
+        } else {
+            ir::types::I32
+        };
+        let csrs = abi::callee_saved_registers(&self.shared_flags);
+
+        let mut csr_stack_size = word_size; // Size of RBP to start with
+        for _reg in &csrs {
+            csr_stack_size += word_size;
+        }
+
+        let stack_offset = -(csr_stack_size as i32);
+        let slot = ir::StackSlotData {
+            kind: ir::StackSlotKind::IncomingArg,
+            size: csr_stack_size,
+            offset: stack_offset,
+        };
+        func.create_stack_slot(slot);
```

Comment:
This is good, but I'd add a comment explaining what is going on.

Stack frame math is hard, but I think the offset is wrong here. It should be relative to the caller's outgoing stack pointer, not your incoming stack pointer. The difference is the return address pushed by the call instruction itself:

```
outgoing sp --> xxxx in caller's frame
incoming sp --> return address
                pushed %rbp
                ...
                last pushed CSR
```

Reply:
I am struggling with this a bit. If I don't put in a stack slot, but I do have local variables, when I run it through Cretonne the offset of that `local` slot is set to 0. Which would seem, to me, to imply that the `incoming_arg` stack slot I make here should have an offset of 0, and I should let `layout_stack` move the locals to `-csr_stack_size`.

But I feel like I'm missing something here, because I see the comment in `stacklayout.rs` which says:
```
    /// Offset of stack slot relative to the stack pointer in the caller.
    ///
    /// On Intel ISAs, the base address is the stack pointer *before* the return address was
    /// pushed. On RISC ISAs, the base address is the value of the stack pointer on entry to the
    /// function.
```

And that would seem to me, to imply that the `locals` slot (assuming no `incoming_arg` slot) should always start at -8. What am I missing?

Reply:
If you don't add any stack slots, the offset of your local slot should be the negative of its size. Is your local slot zero-sized?

The stack layout algorithm does not know that Intel calls automatically save the return address at incoming offset -8, so it is your responsibility to tell it by adding an incoming stack slot at that offset. I think you can add a single incoming stack slot with a size that covers return address + saved fp + saved csrs. It's offset should be the negative of its size, which means it ends exactly at offset 0 where the caller has placed outgoing arguments.

Note that the stack layout algorithm assigns descending offsets (going towards negative infinity), so the size of a local stack slot will affect which offset it gets.

But the offset of a stack slot refers to the lowest address in the stack slot, just like you would do with a pointer to a struct.

Reply:
If you have a single 8-byte local variable, this is what I would expect if you don't add any incoming stack slots:

```
    ss0 = local 8, offset -8
```

This is wrong because that local actually occupies the same bytes as the return address. If you mark these bytes as reserved by adding an incoming stack slot, I would expect you get this:

```
    ss0 = local 8, offset -16
    ss1 = incoming 8, offset -8
```

If you want to reserve space for the return address, a pushed fp and two CSRs, you should get:

```
    ss0 = local 8, offset -40
    ss1 = incoming 32, offset -32
```

Reply:
Ok! I think I finally get it. Thanks for your patience. Latest commit should take care of it.